vault/identity_store_storage_util.go

```diff
@@ -0,0 +1,412 @@
+}
+
+// configStorageEntry contains the properties of packer that needs to survive
+// reboot cycles
+type configStorageEntry struct {
+	NumBuckets int    `json:"num_buckets" structs:"num_buckets" mapstructure:"num_buckets"`
+	ViewPrefix string `json:"view_prefix" structs:"view_prefix" mapstructure:"view_prefix"`
+}
+
+// storageBucketEntry represents a bucket which holds many storage entries
+type storageBucketEntry struct {
+	Key   string                `json:"key" structs:"key" mapstructure:"key"`
+	Items []*entityStorageEntry `json:"items" structs:"items" mapstructure:"items"`
+}
+
+// View returns the storage view configured to be used by the packer
+func (s *storagePacker) View() logical.Storage {
+	return s.view
+}
+
+// Get returns a bucket entry for a given bucket entry key
+func (s *storagePacker) Get(bucketEntryKey string) (*storageBucketEntry, error) {
+	if bucketEntryKey == "" {
+		return nil, fmt.Errorf("missing bucket entry key")
+	}
+
+	// When trying to load entities during startup, scanning the view will
+	// attempt a read on the packer config file as a bucket entry; ignore it.
+	// There won't be a bucket entry with this same prefix as it is a reserved
+	// keyword.
+	if strings.HasPrefix(bucketEntryKey, packerConfigLocation) {
+		return nil, nil
+	}
+
+	// Read from the underlying view
+	storageEntry, err := s.view.Get(bucketEntryKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read packed storage entry: %v", err)
+	}
+	if storageEntry == nil {
+		return nil, nil
+	}
+
+	// Decompress the stored value
+	decompressedBucketEntryBytes, err := snappy.Decode(nil, storageEntry.Value)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decompress packed storage entry: %v", err)
+	}
+
+	// JSON decode it
+	bucketEntry := &storageBucketEntry{}
+	err = jsonutil.DecodeJSON(decompressedBucketEntryBytes, bucketEntry)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode packed storage entry: %v", err)
+	}
+
+	return bucketEntry, nil
+}
+
+// upsert either inserts a new entry to the bucket or updates an existing one
+// if an entry with a matching key is already present.
+func (s *storageBucketEntry) upsert(entry *entityStorageEntry) error {
+	if s == nil {
+		return fmt.Errorf("nil storage bucket entry")
+	}
+
+	if entry == nil {
+		return fmt.Errorf("nil entry")
+	}
+
+	if entry.ID == "" {
+		return fmt.Errorf("missing entity ID")
+	}
+
+	// Look for an entry with matching key and don't modify the collection
+	// while iterating
+	foundIdx := -1
+	for itemIdx, item := range s.Items {
+		if item.ID == entry.ID {
+			foundIdx = itemIdx
+			break
+		}
+	}
+
+	// If there is no match, append the entry, otherwise update it
+	if foundIdx == -1 {
+		s.Items = append(s.Items, entry)
+	} else {
+		s.Items[foundIdx] = entry
+	}
+
+	return nil
+}
+
+// BucketIndex returns the bucket key index for a given storage entry key
+func (s *storagePacker) BucketIndex(key string) uint8 {
```

Comment:
Good question. I was thinking of making the hash function type configurable. On second thoughts, there is not much value is giving that option. Even if we did, we could just have the type of the hash function to be used and not depend on a single instance of one. Locking around it also seems not necessary. I'll remove the locks and create a new md5 every time.